### PR TITLE
eos: Add com.endlessm.EknServices2 to proxy app

### PR DESCRIPTION
**Pull request by ptomato**
All of the possible versions of EknServices should be hidden by the
"Endless Platform" proxy app.

https://phabricator.endlessm.com/T19806

### CODE REVIEW

File: plugins/flatpak/gs-flatpak.c [hunk 1]
```diff
@@ -1868,25 +1868,18 @@ gs_plugin_refine_item_state (GsFlatpak *self,
 }
 
 static GsApp *
-gs_flatpak_create_runtime (GsPlugin *plugin, GsApp *parent, const gchar *runtime)
+gs_flatpak_create_app_from_info (GsPlugin *plugin, const gchar *source,
```
Comment (cosimoc) [inline plugins/flatpak/gs-flatpak.c]: Would be nice to split this refactor into a separate commit

File: plugins/flatpak/gs-flatpak.c [hunk 2]
```diff
@@ -2618,11 +2629,60 @@ gs_flatpak_refine_runtime_for_install (GsFlatpak *self,
...
+	g_autofree gchar *source = NULL;
+
+	runtime_id = gs_app_get_id (runtime);
+	runtime_branch = gs_app_get_branch (runtime);
+
+	if ((g_strcmp0 (runtime_id, "com.endlessm.Platform") == 0 &&
+	     g_strcmp0 (runtime_branch, "eos3.1") == 0) ||
+	    (g_strcmp0 (runtime_id, "com.endlessm.apps.Platform") == 0 &&
+	     g_strcmp0 (runtime_branch, "1") == 0)) {
+		services_id = g_strdup("com.endlessm.EknServices");
```
Comment (cosimoc) [inline plugins/flatpak/gs-flatpak.c]: Missing space before paren

File: plugins/flatpak/gs-flatpak.c [hunk 3]
```diff
@@ -2618,11 +2629,60 @@ gs_flatpak_refine_runtime_for_install (GsFlatpak *self,
...
+		services_branch = "eos3";
+	} else if (g_strcmp0 (runtime_id, "com.endlessm.apps.Platform") == 0) {
+		services_id = g_strdup_printf("com.endlessm.EknServices%s", runtime_branch);
```
Comment (cosimoc) [inline plugins/flatpak/gs-flatpak.c]: Missing space before paren

File: plugins/flatpak/gs-flatpak.c [hunk 4]
```diff
@@ -2567,6 +2567,72 @@ gs_flatpak_related_should_download (GsFlatpak *self, GsApp *app, FlatpakRelatedR
...
+				           flatpak_ref_get_name (FLATPAK_REF (xref)),
+				           error->message);
+				g_clear_error (&error);
+				continue;
+			}
+
+			/* Special case for EknServices 1 and eosplatform 3.1 */
+			if (g_strcmp0 (services_name_ver, "1") == 0 &&
+			    g_strcmp0 (gs_app_get_id (app), "com.endlessm.Platform") == 0 &&
+			    g_strcmp0 (gs_app_get_branch (app), "eos3.1")) {
```
Comment (cosimoc) [inline plugins/flatpak/gs-flatpak.c]: Should be `g_strcmp0 (...) == 0`, no? (If not, prefer the explicit `!= 0` notation.)

File: plugins/flatpak/gs-flatpak.c [hunk 5]
```diff
@@ -2618,11 +2629,60 @@ gs_flatpak_refine_runtime_for_install (GsFlatpak *self,
...
+{
+	const gchar *runtime_id, *runtime_branch, *services_branch, *arch;
+	g_autofree gchar *services_id = NULL;
+	g_autofree gchar *source = NULL;
+
+	runtime_id = gs_app_get_id (runtime);
+	runtime_branch = gs_app_get_branch (runtime);
+
+	if ((g_strcmp0 (runtime_id, "com.endlessm.Platform") == 0 &&
+	     g_strcmp0 (runtime_branch, "eos3.1") == 0) ||
```
Comment (joaquimrocha) [inline plugins/flatpak/gs-flatpak.c]: Shouldn't it be also for _eos3.2_?

File: plugins/flatpak/gs-flatpak.c [hunk 6]
```diff
@@ -2618,11 +2629,60 @@ gs_flatpak_refine_runtime_for_install (GsFlatpak *self,
...
+		services_branch = "stable";
+	} else {
+		/* Runtime doesn't require an EknServices app */
+		return NULL;
+	}
+
+	/* Construct a GsApp for the EknServices we determined we needed */
+	arch = gs_flatpak_app_get_ref_arch (runtime);
+	source = g_strdup_printf ("app/%s/%s/%s", services_id, arch, services_branch);
+	return gs_flatpak_create_app_from_info (plugin, source, services_id, arch,
```
Comment (joaquimrocha) [inline plugins/flatpak/gs-flatpak.c]: I don't think the app should be created this way. You should rather use `gs_flatpak_create_app` that takes a `FlatpakRef`. That way you also don't have to split up the runtime creation function into yet another app creation one.

File: plugins/flatpak/gs-flatpak.c [hunk 7]
```diff
@@ -2618,11 +2629,60 @@ gs_flatpak_refine_runtime_for_install (GsFlatpak *self,
...
+	                                        services_branch, AS_APP_KIND_GENERIC,
+	                                        FLATPAK_REF_KIND_APP);
+}
+
+static void
+gs_flatpak_add_app_to_list_if_not_installed (GsApp *app, GsAppList *list,
+					     const GHashTable *hash_installed)
+{
+	g_autofree gchar *ref_display = NULL;
+	ref_display = gs_flatpak_app_get_ref_display (app);
```
Comment (joaquimrocha) [inline plugins/flatpak/gs-flatpak.c]: You could make these into the same line where you declare it.

File: plugins/flatpak/gs-flatpak.c [hunk 8]
```diff
@@ -2648,19 +2708,13 @@ gs_flatpak_get_list_for_install (GsFlatpak *self, GsApp *app,
...
-			gs_app_list_add (list, runtime);
-		}
-	}
+	if (runtime != NULL)
+		gs_flatpak_add_app_to_list_if_not_installed (runtime, list, hash_installed);
+
+	/* add services flatpak */
+	services = gs_flatpak_get_services_app_for_runtime (self->plugin, runtime);
+	if (services != NULL)
+		gs_flatpak_add_app_to_list_if_not_installed (services, list, hash_installed);
```
Comment (joaquimrocha) [inline plugins/flatpak/gs-flatpak.c]: I understand the consistence with the code above and not repeating two blocks. However, this involves more changes to `gs-flatpak` than I really wanted. In this case, the code you're abstracting `gs_flatpak_add_app_to_list_if_not_installed` is only worth it because of the comments, but the comments in the function look out of place to me (they're referring to the installation). So just repeat the code, or maybe do use the function but return a boolean and use it to print the debug messages from outside.

File: plugins/flatpak/gs-flatpak.c [hunk 9]
```diff
@@ -2567,6 +2567,72 @@ gs_flatpak_related_should_download (GsFlatpak *self, GsApp *app, FlatpakRelatedR
...
+
+			/* FIXME: can you compare pointer equality like this
+			 * because of the cache? */
+			if (runtime == app)
+				found_app_that_uses_runtime = TRUE;
+		}
+
+		return found_app_that_uses_runtime;
+	}
+
```
Comment (joaquimrocha) [inline plugins/flatpak/gs-flatpak.c]: I am afraid this will have an impact in performance: 1) Listing all installed refs with libflatpak is slow; 2) IO per each of those apps ditto.
I know this should happen only during installation/update, but still this code seems like a lot of complexity added "just" for skipping EknServices.

I understand that keeping a number of EknServices around may have an impact on the search, but I would rather include it's clean up as part of cleaning up unneeded legacy runtimes. So when we get rid of the eos3.2 Platform runtime, we should also uninstall EknServices (see `...get_list_for_remove`).

File: plugins/flatpak/gs-flatpak.c (hunk 5 above)
Comment (ptomato) [reply plugins/flatpak/gs-flatpak.c]: I don't think that's necessary, there are no knowledge apps published that use the `com.endlessm.Platform//eos3.2` runtime (and never will be.)

File: plugins/flatpak/gs-flatpak.c (hunk 9 above)
Comment (ptomato) [reply plugins/flatpak/gs-flatpak.c]: OK, I agree with what you say and will change it.

However, the situation that I was trying to avoid was this:
- user doesn't have any SDK2 apps installed
- user updates Endless Platform
- com.endlessm.EknServices2 is in the list
- com.endlessm.EknServices2 gets installed as part of the Endless Platform update
- com.endlessm.EknServices2 pulls in SDK2 into the update

Do you know if that situation will actually happen, and/or what's a better way to avoid it?

File: plugins/flatpak/gs-flatpak.c (hunk 8 above)
Comment (ptomato) [reply plugins/flatpak/gs-flatpak.c]: I'm not sure I understand what you're asking; I can't figure out which comments you're referring to.

File: plugins/flatpak/gs-flatpak.c [hunk 10]
```diff
@@ -2500,6 +2539,15 @@ gs_flatpak_get_list_for_remove (GsFlatpak *self, GsApp *app,
 		return FALSE;
 	}
 
+	/* If this is the SDK runtime, remove the associated EknServices.
+	 * Ignores EknServices associated with com.endlessm.Platform//eos3.1. */
+	if (g_strcmp0 (ref_name, "com.endlessm.apps.Platform") == 0) {
+		g_autoptr(GsApp) services = gs_flatpak_get_services_app_for_runtime (self, app);
+		if (services != NULL) {
```
Comment (cosimoc) [inline plugins/flatpak/gs-flatpak.c]: Please remove curly braces for single-line blocks

File: plugins/flatpak/gs-flatpak.c [hunk 11]
```diff
@@ -2647,19 +2711,13 @@ gs_flatpak_get_list_for_install (GsFlatpak *self, GsApp *app,
...
-				 gs_flatpak_app_get_ref_name (runtime),
-				 gs_flatpak_app_get_ref_branch (runtime));
-			gs_app_list_add (list, runtime);
-		}
-	}
+	if (runtime != NULL)
+		gs_flatpak_add_app_to_list_if_not_installed (runtime, list, hash_installed);
+
+	/* add services flatpak */
+	services = gs_flatpak_get_services_app_for_runtime (self, runtime);
```
Comment (cosimoc) [inline plugins/flatpak/gs-flatpak.c]: Does `services` (or even `runtime`) not get leaked?

File: plugins/flatpak/gs-flatpak.c [hunk 12]
```diff
@@ -2477,18 +2477,57 @@ gs_flatpak_app_remove_source (GsFlatpak *self,
...
+	g_autofree gchar *services_id = NULL;
+	g_autofree gchar *description = NULL;
+	g_autoptr(FlatpakRef) services_ref = NULL;
+	g_autoptr(GError) error = NULL;
+
+	runtime_id = gs_app_get_id (runtime);
+	runtime_branch = gs_app_get_branch (runtime);
+
+	if ((g_strcmp0 (runtime_id, "com.endlessm.Platform") == 0 &&
+	     g_strcmp0 (runtime_branch, "eos3.1") == 0) ||
```
Comment (cosimoc) [inline plugins/flatpak/gs-flatpak.c]: Does `com.endlessm.Platform` version `eos3.2` not require `EknServices` version `eos3`?

Comment (ptomato) [reply plugins/flatpak/gs-flatpak.c]: No, there are no knowledge apps that use `com.endlessm.Platform//eos3.2` and never will be.

File: plugins/flatpak/gs-flatpak.c (hunk 11 above)
Comment (ptomato) [reply plugins/flatpak/gs-flatpak.c]: I don't think so, I believe the `GsAppList` takes ownership.

File: plugins/flatpak/gs-flatpak.c (hunk 12 above)
Comment (cosimoc) [reply plugins/flatpak/gs-flatpak.c]: Sorry, did not see the previous similar comment from @joaquimrocha. Sounds good then.

File: plugins/flatpak/gs-flatpak.c (hunk 11 above)
Comment (cosimoc) [reply plugins/flatpak/gs-flatpak.c]: `gs_app_list_add()` adds a new ref as far as I can see.

Comment (ptomato) [reply plugins/flatpak/gs-flatpak.c]: `runtime` is not leaked (comes from a transfer none function), but the `services` leak is fixed now.

**Comment by cosimoc** [comment]
LGTM now but I will leave unmerged for @joaquimrocha to take another look too.

**Comment by ptomato** [comment]
I would also prefer not to merge this until I've figured out how to test it.

File: plugins/flatpak/gs-flatpak.c [hunk 13]
```diff
@@ -2617,6 +2664,22 @@ gs_flatpak_refine_runtime_for_install (GsFlatpak *self,
...
 }
 
+static void
+gs_flatpak_add_app_to_list_if_not_installed (GsApp *app, GsAppList *list,
+					     const GHashTable *hash_installed)
+{
+	g_autofree gchar *ref_display = gs_flatpak_app_get_ref_display (app);
+	if (g_hash_table_contains (hash_installed, ref_display)) {
+		g_debug ("%s is already installed, so skipping",
+			 gs_app_get_id (app));
```
Comment (joaquimrocha) [inline plugins/flatpak/gs-flatpak.c]: You could do an early return here to skip the else block.

File: plugins/flatpak/gs-flatpak.c (hunk 8 above)
Comment (ptomato) [reply plugins/flatpak/gs-flatpak.c]: OK, I tried repeating the code but it looks wrong to me... I still get the feeling I am misinterpreting your comment and looking at the wrong thing, but if you don't feel strongly about it then we can just drop it.

File: plugins/flatpak/gs-flatpak.c (hunk 13 above)
Comment (ptomato) [reply plugins/flatpak/gs-flatpak.c]: Fixed.

File: plugins/flatpak/gs-flatpak.c [hunk 14]
```diff
@@ -2477,18 +2477,82 @@ gs_flatpak_app_remove_source (GsFlatpak *self,
...
+	description = g_strdup_printf ("app/%s/%s/%s", services_id, arch, services_branch);
+	services_ref = flatpak_ref_parse (description, &error);
+	if (!services_ref) {
+		g_critical ("Invalid flatpak ref description %s: %s", description,
+		            error->message);
+		return NULL;
+	}
+
+	retval = gs_flatpak_create_app (self, services_ref);
+	gs_app_set_origin (retval, gs_app_get_origin (runtime));
```
Comment (joaquimrocha) [inline plugins/flatpak/gs-flatpak.c]: If you do this then you'll make it look for the services app in the runtime's remote, and this is why you're getting the failure you've reported in [here](https://phabricator.endlessm.com/T19806#504381). Also, by setting the origin (and if the origin is not _NULL_), the next line will do nothing as it first verifies if an origin is already set before it proceeds to look for it.
So maybe all that's needed is to delete this line.

File: plugins/flatpak/gs-flatpak.c [hunk 15]
```diff
@@ -2477,18 +2477,82 @@ gs_flatpak_app_remove_source (GsFlatpak *self,
...
+static GsApp *
+gs_flatpak_get_services_app_for_runtime (GsFlatpak *self, GsApp *runtime,
+                                         GCancellable *cancellable)
+{
+	const gchar *runtime_id, *runtime_branch, *services_branch, *arch;
+	g_autofree gchar *services_id = NULL;
+	g_autofree gchar *description = NULL;
+	g_autoptr(FlatpakRef) services_ref = NULL;
+	g_autoptr(GError) error = NULL;
+	g_autoptr(GsApp) retval = NULL;
```
Comment (joaquimrocha) [inline plugins/flatpak/gs-flatpak.c]: Let's call this `app` or `services_app` instead.

File: plugins/flatpak/gs-flatpak.c [hunk 16]
```diff
@@ -2477,18 +2477,82 @@ gs_flatpak_app_remove_source (GsFlatpak *self,
...
+		/* Runtime doesn't require an EknServices app */
+		return NULL;
+	}
+
+	/* Construct a GsApp for the EknServices we determined we needed */
+	arch = gs_flatpak_app_get_ref_arch (runtime);
+	description = g_strdup_printf ("app/%s/%s/%s", services_id, arch, services_branch);
+	services_ref = flatpak_ref_parse (description, &error);
+	if (!services_ref) {
+		g_critical ("Invalid flatpak ref description %s: %s", description,
```
Comment (joaquimrocha) [inline plugins/flatpak/gs-flatpak.c]: I am afraid that due to a mistake we get it to abort here, rather than "simply" not getting the services app installed.
So I'd make this a warning instead... I know the way I am proposing makes it harder to notice when this happens but still, having no _EknServices_ and having users reporting that they cannot search in apps seems better than eventually braking the installation of apps.

File: plugins/flatpak/gs-flatpak.c [hunk 17]
```diff
@@ -2477,18 +2477,82 @@ gs_flatpak_app_remove_source (GsFlatpak *self,
...
+	} else {
+		/* Runtime doesn't require an EknServices app */
+		return NULL;
+	}
+
+	/* Construct a GsApp for the EknServices we determined we needed */
+	arch = gs_flatpak_app_get_ref_arch (runtime);
+	description = g_strdup_printf ("app/%s/%s/%s", services_id, arch, services_branch);
+	services_ref = flatpak_ref_parse (description, &error);
+	if (!services_ref) {
```
Comment (joaquimrocha) [inline plugins/flatpak/gs-flatpak.c]: Richard doesn't like the `!pointer` notation (prefers `!= NULL`). So even though I prefer to do it as you did, I try to keep the coding style in core modules close to upstream (even if this will never be merged). Small nitpick though.

File: plugins/flatpak/gs-flatpak.c [hunk 18]
```diff
@@ -2477,18 +2477,82 @@ gs_flatpak_app_remove_source (GsFlatpak *self,
...
+ * Returns NULL if no services app was required or there was an error.
+ * Does not take a GError since any error should be recoverable. However, in
+ * order to make sure the operation was not cancelled, you should check the
+ * status of the GCancellable after calling this.
+ */
+static GsApp *
+gs_flatpak_get_services_app_for_runtime (GsFlatpak *self, GsApp *runtime,
+                                         GCancellable *cancellable)
+{
+	const gchar *runtime_id, *runtime_branch, *services_branch, *arch;
```
Comment (joaquimrocha) [inline plugins/flatpak/gs-flatpak.c]: Usually we don't have declarations like this in g-s. Even if I prefer these actually.
I am also not sure if the C99 standard will complain or not, but I'd make these several lines. (Just a nitpick though, not a blocker for me, unless there's a compiler warning)

File: plugins/flatpak/gs-flatpak.c [hunk 19]
```diff
@@ -2500,6 +2564,19 @@ gs_flatpak_get_list_for_remove (GsFlatpak *self, GsApp *app,
 		return FALSE;
 	}
 
+	/* If this is the SDK runtime, remove the associated EknServices.
+	 * Ignores EknServices associated with com.endlessm.Platform//eos3.1. */
+	if (g_strcmp0 (ref_name, "com.endlessm.apps.Platform") == 0) {
```
Comment (joaquimrocha) [inline plugins/flatpak/gs-flatpak.c]: As you pointed out recently, since we never remove a runtime from g-s. Even if we have to do so in the future, it's probably a good idea not to carry this code for now (because e.g. if it gets outdated, then when we do implement the removal of runtimes we could forget to update or verify this as long as it compiles...).

File: plugins/flatpak/gs-flatpak.c [hunk 20]
```diff
@@ -2647,19 +2742,18 @@ gs_flatpak_get_list_for_install (GsFlatpak *self, GsApp *app,
...
-				 gs_flatpak_app_get_ref_name (runtime),
-				 gs_flatpak_app_get_ref_branch (runtime));
-			gs_app_list_add (list, runtime);
-		}
+	if (runtime != NULL)
+		gs_flatpak_add_app_to_list_if_not_installed (runtime, list, hash_installed);
+
+	/* add services flatpak */
+	services = gs_flatpak_get_services_app_for_runtime (self, runtime, cancellable);
+	if (g_cancellable_is_cancelled (cancellable)) {
```
Comment (joaquimrocha) [inline plugins/flatpak/gs-flatpak.c]: You can use `g_cancellable_set_error_if_cancelled` which sets the error and returns `TRUE` if it's been cancelled. So all you'd need to do is return FALSE in that case.

File: plugins/flatpak/gs-flatpak.c (hunk 14 above)
Comment (ptomato) [reply plugins/flatpak/gs-flatpak.c]: OMG, I stared at this for so long and didn't notice that I had failed to remove this line! Thanks!

File: plugins/flatpak/gs-flatpak.c (hunk 18 above)
Comment (ptomato) [reply plugins/flatpak/gs-flatpak.c]: Multiple declarations per line have been a thing in C since at least [1978](https://archive.org/details/TheCProgrammingLanguageFirstEdition) :smile: But I agree it's better to keep the style of the surrounding code.

File: plugins/flatpak/gs-flatpak.c [hunk 21]
```diff
@@ -2647,19 +2731,22 @@ gs_flatpak_get_list_for_install (GsFlatpak *self, GsApp *app,
...
+
+	/* add services flatpak */
+	services_app = gs_flatpak_get_services_app_for_runtime (self, runtime, cancellable);
+	if (g_cancellable_set_error_if_cancelled (cancellable, error))
+		return FALSE;
+	if (services_app != NULL)
+		gs_flatpak_add_app_to_list_if_not_installed (services_app, list, hash_installed);
+
+	/* add services flatpak's runtime, if different from app's runtime */
+	if (!gs_flatpak_refine_runtime_for_install (self, services_app, cancellable, error))
```
Comment (joaquimrocha) [inline plugins/flatpak/gs-flatpak.c]: I think we will have problems here (and below) when `services_app` is NULL. This should be part of the `if` block above.

**Comment by ptomato** [comment]
Fixed the last comment.
